api/src/main/java/org/pmiops/workbench/access/AccessModuleServiceImpl.java

```diff
@@ -0,0 +1,133 @@
+  private static final Logger logger = Logger.getLogger(AccessModuleServiceImpl.class.getName());
+
+  private final Provider<List<DbAccessModule>> dbAccessModulesProvider;
+  private final Clock clock;
+
+  private final UserAccessModuleDao userAccessModuleDao;
+  private final UserDao userDao;
+  private final UserServiceAuditor userServiceAuditor;
+  private final Provider<WorkbenchConfig> configProvider;
+
+  private static final BiMap<AccessModule, AccessModuleName> CLIENT_TO_STORAGE_ACCESS_MODULE =
+      ImmutableBiMap.<AccessModule, AccessModuleName>builder()
+          .put(AccessModule.TWO_FACTOR_AUTH, AccessModuleName.TWO_FACTOR_AUTH)
+          .put(AccessModule.ERA_COMMONS, AccessModuleName.ERA_COMMONS)
+          .put(AccessModule.COMPLIANCE_TRAINING, AccessModuleName.RT_COMPLIANCE_TRAINING)
+          .put(AccessModule.RAS_LINK_LOGIN_GOV, AccessModuleName.RAS_LOGIN_GOV)
+          .put(AccessModule.DATA_USE_AGREEMENT, AccessModuleName.DATA_USER_CODE_OF_CONDUCT)
+          .put(AccessModule.PUBLICATION_CONFIRMATION, AccessModuleName.PUBLICATION_CONFIRMATION)
+          .put(AccessModule.PROFILE_CONFIRMATION, AccessModuleName.PROFILE_CONFIRMATION)
+          .build();
+
+  private static final BiMap<BypassTimeTargetProperty, AccessModuleName>
+      AUDIT_TO_STORAGE_ACCESS_MODULE =
+          ImmutableBiMap.<BypassTimeTargetProperty, AccessModuleName>builder()
+              .put(BypassTimeTargetProperty.ERA_COMMONS_BYPASS_TIME, AccessModuleName.ERA_COMMONS)
+              .put(
+                  BypassTimeTargetProperty.COMPLIANCE_TRAINING_BYPASS_TIME,
+                  AccessModuleName.RT_COMPLIANCE_TRAINING)
+              .put(
+                  BypassTimeTargetProperty.TWO_FACTOR_AUTH_BYPASS_TIME,
+                  AccessModuleName.TWO_FACTOR_AUTH)
+              .put(BypassTimeTargetProperty.RAS_LINK_LOGIN_GOV, AccessModuleName.RAS_LOGIN_GOV)
+              .put(
+                  BypassTimeTargetProperty.DATA_USE_AGREEMENT_BYPASS_TIME,
+                  AccessModuleName.DATA_USER_CODE_OF_CONDUCT)
+              .build();
+
+  @Autowired
+  public AccessModuleServiceImpl(
+      Provider<List<DbAccessModule>> dbAccessModulesProvider,
+      Clock clock,
+      UserAccessModuleDao userAccessModuleDao,
+      UserDao userDao,
+      UserServiceAuditor userServiceAuditor,
+      Provider<WorkbenchConfig> configProvider) {
+    this.dbAccessModulesProvider = dbAccessModulesProvider;
+    this.clock = clock;
+    this.userAccessModuleDao = userAccessModuleDao;
+    this.userDao = userDao;
+    this.userServiceAuditor = userServiceAuditor;
+    this.configProvider = configProvider;
+  }
+
+  @Override
+  public void updateBypassTime(long userId, AccessModule accessModuleName, boolean isBypassed) {
+    final DbUser user = userDao.findUserByUserId(userId);
+    final List<DbUserAccessModule> dbUserAccessModules = userAccessModuleDao.getAllByUser(user);
+    DbAccessModule accessModule =
```

Comment:
Since this failure can happen regardless of what's in the other DB tables, we should move this to the top of updateBypassTime().  We'll likely need to reuse this so I'd also prefer to break this out into a small helper method.

Reply:
Moved to the top of the method. Which part you think will be reusable? I assume bypassable check just happens here. But we can pull stuff out if we have other use cases?

Reply:
Thanks.  I mean something very simple like
```
DbAccessModule getOrThrow(AccessModule accessModuleName)
```

Reply:
ah I see. Done!